crypto/dsa/dsa_aid.c

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+
+#include <openssl/objects.h>
+#include "crypto/dsa.h"
+
+#define ASN1_SEQUENCE 0x30
+#define ASN1_OID 0x06
+
+/* dsaWithSHA OIDs are of the form: (1 3 14 3 2 |n|) */
+#define ENCODE_ALGORITHMIDENTIFIER_SHA(name, n)                         \
+    static const unsigned char algorithmidentifier_##name##_der[] = {   \
+        ASN1_SEQUENCE, 0x06,                                            \
+          ASN1_OID, 0x09, 1 * 40 + 3, 14, 3, n,                         \
+}
+
+ENCODE_ALGORITHMIDENTIFIER_SHA(sha, 13);
+ENCODE_ALGORITHMIDENTIFIER_SHA(sha1, 27);
+
+/* dsaWithSHA OIDs are of the form: (2 16 840 1 101 3 4 3 |n|) */
+#define ENCODE_ALGORITHMIDENTIFIER_SHA2(name, n)                         \
+    static const unsigned char algorithmidentifier_##name##_der[] = {   \
+        ASN1_SEQUENCE, 0x0b,                                            \
+          ASN1_OID, 0x09, 2 * 40 + 16, 0x86, 0x48, 1, 101, 3, 4, 3, n   \
+}
+
+ENCODE_ALGORITHMIDENTIFIER_SHA2(sha224, 1);
+ENCODE_ALGORITHMIDENTIFIER_SHA2(sha256, 2);
+
+#define MD_CASE(name)                                                   \
+    case NID_##name:                                                    \
+        *len = sizeof(algorithmidentifier_##name##_der);                \
+        return algorithmidentifier_##name##_der
+
+const unsigned char *dsa_algorithmidentifier_encoding(int md_nid, size_t *len)
+{
+    switch (md_nid) {
```

Comment:
All SHA variants need to be in this table.. (including the truncated ones)
i.e...
Federal Information Processing Standard (FIPS) 186-4, Digital Signature Standard (DSS), specifies three NIST-approved digital signature algorithms: DSA, RSA, and ECDSA. All three are used to generate and verify digital signatures, in conjunction with an approved hash function specified in FIPS 180-4, Secure Hash Standard or FIPS 202, SHA-3 Standard: Permutation-Based Hash and Extendable-Output Functions.

Reply:
Hold on a minute...  This table here is about plain DSA.  Not RSA, not ECDSA.  Similar tables should and will be added in the RSA and ECDSA source (I do this in a sneaky way for RSA in #10557)

So for *this* table, the interesting OIDs are DSAWith{name}, {name}WithDSA, that sort of thing.  I haven't been able to find any that has SHA-3.  It's possible that I've starred myself blind.  Would you mind helping out?

Reply:
I cant find the OID's for the truncated ones (so they might not exist), but here are the others..
If we are supporting SHA3 it should be supported also for signing.
https://csrc.nist.gov/projects/computer-security-objects-register/algorithm-registration

Reply:
Thanks for the link.  Will add.

Reply:
Fixed.